data_structures/binary_search_tree.ts

```diff
@@ -0,0 +1,199 @@
+ *
+ * @template T The data type of the values in the binary tree.
+ */
+export class BinarySearchTree<T> {
+  rootNode?: TreeNode<T>;
+
+  /**
+   * Instantiates the binary search tree.
+   *
+   * @param rootNode The root node.
+   */
+  constructor() {
+    this.rootNode = undefined;
+  }
+
+  /**
+   * Checks, if the binary search tree is empty, i. e. has no root node.
+   *
+   * @returns Whether the binary search tree is empty.
+   */
+  isEmpty(): boolean {
+    return this.rootNode === undefined;
+  }
+
+  /**
+   * Searches for data in the current binary search tree.
+   *
+   * @param data The data to search for.
+   */
+  search(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      throw new Error('No root node defined.');
+    }
+
+    let currentNode = this.rootNode;
+    while (currentNode.data !== data) {
+      if (data > currentNode.data) {
+        if (!currentNode.rightChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.rightChild;
+      } else {
+        if (!currentNode.leftChild) {
+          throw new Error('No such element found.');
+        }
+
+        currentNode = currentNode.leftChild;
+      }
+    }
+
+    return currentNode;
+  }
+
+  /**
+   * Inserts the given data into the binary search tree.
+   *
+   * @param data The data to be stored in the binary search tree.
+   * @returns
+   */
+  insert(data: T): TreeNode<T> {
+    if (!this.rootNode) {
+      this.rootNode = new TreeNode<T>(data);
+      return this.rootNode;
+    }
+
+    let currentNode: TreeNode<T> = this.rootNode;
+    while (true) {
+      if (data > currentNode.data) {
+        if (currentNode.rightChild) {
+          currentNode = currentNode.rightChild;
+        } else {
+          currentNode.rightChild = new TreeNode<T>(data);
+          return currentNode.rightChild;
+        }
+      } else {
+        if (currentNode.leftChild) {
+          currentNode = currentNode.leftChild;
+        } else {
+          currentNode.leftChild = new TreeNode<T>(data);
+          return currentNode.leftChild;
+        }
+      }
+    }
+  }
+
+  /**
+   * Finds the minimum value of the binary search tree.
+   *
+   * @returns The minimum value of the binary search tree
+   */
+  findMin(): T {
+    if (!this.rootNode) {
+      throw new Error('No root node defined');
+    }
+
+    const traverse = (node: TreeNode<T>): T => {
+      return !node.leftChild ? node.data : traverse(node.leftChild);
+    };
+
+    return traverse(this.rootNode);
+  }
+
+  /**
+   * Finds the maximum value of the binary search tree.
+   *
+   * @returns The maximum value of the binary search tree
+   */
+  findMax(): T {
+    if (!this.rootNode) {
+      throw new Error('No root node defined');
+    }
+
+    const traverse = (node: TreeNode<T>): T => {
+      return !node.rightChild ? node.data : traverse(node.rightChild);
```

Comment:
You could use a loop instead. Matter of taste, feel free to resolve this.

Reply:
As I used this type of closured also for the other traverse methods now, I will resolve this.